### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by bojanv**
## Purpose

Major refactoring and addressing the MsalUiException error via decorators (ExceptionFilter) and calling the refresh token with the use of HttpClientFactory for resilient Http request to GraphAPI.

Reference to the issue: https://github.com/Azure-Samples/active-directory-b2c-dotnet-webapp-and-webapi/issues/9

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[x] Bugfix
[ ] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone https://github.com/snobu/active-directory-aspnetcore-webapp-openidconnect-v2.git
cdactive-directory-aspnetcore-webapp-openidconnect-v2.git
git checkout refactor
dotnet run

* Test the code
<!-- Add steps to run the tests suite and/or manually test -->
Set details in **appsettings.json** and run via dotnet cli or Visual Studio or Visual Studio Code. Login and  click on Profile link.

## What to Check
Exception Filter: https://github.com/snobu/active-directory-aspnetcore-webapp-openidconnect-v2/blob/refactor/Infrastructure/MsalUiRequiredExceptionFilterAttribute.cs (folder **Infrastructure**)
Fixing the problem with Msal exception type: https://github.com/snobu/active-directory-aspnetcore-webapp-openidconnect-v2/blob/refactor/Controllers/HomeController.cs (Folder **Controllers**, method **Profile**)
GraphApi implementation: https://github.com/snobu/active-directory-aspnetcore-webapp-openidconnect-v2/blob/refactor/Services/GraphApiOperationService.cs (folder **Services**)


## Other Information
Please check additional information as an answer at  https://github.com/Azure-Samples/active-directory-b2c-dotnet-webapp-and-webapi/issues/9

### CODE REVIEW

File: Infrastructure/MsalUiRequiredExceptionFilterAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+{
+    public class MsalUiRequiredExceptionFilterAttribute : ExceptionFilterAttribute
+    {
+        public string[] Scopes { get; set; }
+        
+        public override void OnException(ExceptionContext context)
+        {
+            if (context.Exception is MsalUiRequiredException msalUiRequiredException)
+            {
+                if (CanbeSolvedByReSignInUser(msalUiRequiredException))
```
Comment (bgavrilMS) [inline Infrastructure/MsalUiRequiredExceptionFilterAttribute.cs]: CanBe

File: Infrastructure/AzureAdOptions.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+namespace WebApp_OpenIDConnect_DotNet.Infrastructure
```
Comment (jmprieur) [inline Infrastructure/AzureAdOptions.cs]: This should not be needed. it's already part of ASP.NET Core: See https://github.com/aspnet/AspNetCore/blob/master/src/Azure/AzureAD/Authentication.AzureAD.UI/src/AzureADOptions.cs
why did you add it?

File: Controllers/HomeController.cs [hunk 3]
```diff
@@ -1,139 +1,52 @@
...
-                }
-                else
-                {
-                    throw;
-                }
-            }
-        }
-
-        private static bool CanbeSolvedByReSignInUser(MsalUiRequiredException ex)
+        [MsalUiRequiredExceptionFilter(Scopes = new[] {Constants.ScopeUserRead})]
```
Comment (jmprieur) [inline Controllers/HomeController.cs]: Nice!

File: Extensions/AadIssuerValidator.cs [hunk 4]
```diff
@@ -23,29 +21,29 @@ public static class AadIssuerValidator
         /// <returns>The <c>issuer</c> if it's valid, or otherwise <c>SecurityTokenInvalidIssuerException</c> is thrown</returns>
         public static string ValidateAadIssuer(string issuer, SecurityToken securityToken, TokenValidationParameters validationParameters)
         {
-            JwtSecurityToken jwtToken = securityToken as JwtSecurityToken;
+            var jwtToken = securityToken as JwtSecurityToken;
```
Comment (jmprieur) [inline Extensions/AadIssuerValidator.cs]: not sure what's the point of replacing the types by var?

File: Extensions/ClaimPrincipalExtension.cs [hunk 5]
```diff
@@ -12,16 +12,16 @@ public static class ClaimsPrincipalExtension
         /// <returns>A string corresponding to an account identifier as defined in <see cref="Microsoft.Identity.Client.AccountId.Identifier"/></returns>
         public static string GetMsalAccountId(this ClaimsPrincipal claimsPrincipal)
         {
-            string userObjectId = GetObjectId(claimsPrincipal);
-            string tenantId = GetTenantId(claimsPrincipal);
+            var userObjectId = GetObjectId(claimsPrincipal);
```
Comment (jmprieur) [inline Extensions/ClaimPrincipalExtension.cs]: I even less see the point for strings :)

File: Infrastructure/Bootstrapper.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+using WebApp_OpenIDConnect_DotNet.Interfaces;
+using WebApp_OpenIDConnect_DotNet.Services;
+
+namespace WebApp_OpenIDConnect_DotNet.Infrastructure
+{
+    public static class Bootstrapper
+    {
+        public static void InitializeDefault(this IServiceCollection services, IConfiguration configuration)
```
Comment (jmprieur) [inline Infrastructure/Bootstrapper.cs]: Good point, you are suggesting that the Startup.cs should be even simpler.
I got this feedback already as well.

Any objections if I moved this file and the files under Extensions to a new library?

File: Interfaces/ITokenAcquisition.cs [hunk 7]
```diff
@@ -69,7 +69,7 @@ public interface ITokenAcquisition
         /// };
         /// </code>
         /// </example>
-        void AddAccountToCacheFromJwt(JwtBearer.TokenValidatedContext tokenValidationContext, IEnumerable<string> scopes = null);
```
Comment (jmprieur) [inline Interfaces/ITokenAcquisition.cs]: Warning, there are 2 TokenValidatedContext 
a JwtBearer one and an OIDCOne
I would prefer that this is explicit? is it an issue?

File: Services/GraphApiOperationService.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
+using System;
```
Comment (jmprieur) [inline Services/GraphApiOperationService.cs]: Should we just use the Graph SDK ?

File: Startup.cs [hunk 9]
```diff
@@ -22,116 +15,12 @@ public Startup(IConfiguration configuration)
...
-                    var policy = new AuthorizationPolicyBuilder()
-                        .RequireAuthenticatedUser()
-                        .Build();
-                    options.Filters.Add(new AuthorizeFilter(policy));
-                })
-                .SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
+            //default system actions and options integration with dependency system
+            services.InitializeDefault(Configuration); 
+            //init AzureAd specific configuration
+            services.InitializeAuthentication(Configuration);
```
Comment (jmprieur) [inline Startup.cs]: I like the apparent simplification
The issue here is you don't specify the token cache implementation.

File: Infrastructure/AzureAdOptions.cs (hunk 2 above)
Comment (bojanv) [reply Infrastructure/AzureAdOptions.cs]: I was following convention with [IOptions ](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-2.2)dialog to be able to pull in additional details. Can be removed, but the point here is to have it inserted in different places via dependency injection to get required data from user settings perspective.
```
services.Configure<WebOptions>(configuration);
services.Configure<AzureAdOptions>(configuration.GetSection("AzureAd"));
```

File: Extensions/AadIssuerValidator.cs (hunk 4 above)
Comment (bojanv) [reply Extensions/AadIssuerValidator.cs]: Frankly, it is a coding style I prefer. From code reading perspective you already see, that is JwtSecurityToken due to cast. Compiler will determine the type on the end. It is not required, it is more of a preference (more [inline with .NET framework](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/var) coding styles)

File: Extensions/ClaimPrincipalExtension.cs (hunk 5 above)
Comment (bojanv) [reply Extensions/ClaimPrincipalExtension.cs]: I understand, can be replaced with explicit type. From code perspective, it is the same, from reading and code style perspective, it defers. :)

File: Infrastructure/Bootstrapper.cs (hunk 6 above)
Comment (bojanv) [reply Infrastructure/Bootstrapper.cs]: No objections, feel free to moved it as you wish 👍

File: Interfaces/ITokenAcquisition.cs (hunk 7 above)
Comment (bojanv) [reply Interfaces/ITokenAcquisition.cs]: Valid point, agree.

File: Services/GraphApiOperationService.cs (hunk 8 above)
Comment (bojanv) [reply Services/GraphApiOperationService.cs]: We can, but the point of this service is to have a resilient HTTP request with the use of client factories in ASP.NET Core. More [here](https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests). The point is to use best practices, when performing calls to endpoints in a way that feels native to ASP.NET Core developers to "force" developer to use best practices.

File: Startup.cs (hunk 9 above)
Comment (bojanv) [reply Startup.cs]: True. I can do that as well. I left the implementation of the token cache in the `Bootstrapper.cs` and focus on refactoring bits with re-issuing token and modify it via filter.

I didn't focus on the `TokenAcquisition.cs` class. That can be easily refactored and merge to Boostraper.cs for easier usage and same coding style.

Do it do that as well?

Comment (jmprieur) [reply Startup.cs]: as you prefer @bojanv 
I'm not sure we'd want to merge the TokenAcquisition.cs class (reusable Aspnet service) with Bootstrapper?
how would you surface the choice of the cache implementation? something like .WithSQLTokenCache().

I can also do it when I've merged.

File: Extensions/AadIssuerValidator.cs (hunk 4 above)
Comment (jmprieur) [reply Extensions/AadIssuerValidator.cs]: @bgavrilMS to decide on this ones :)

File: Services/GraphApiOperationService.cs (hunk 8 above)
Comment (jmprieur) [reply Services/GraphApiOperationService.cs]: thanks for the explanation, @bojanv 
this makes sense.

File: Extensions/AadIssuerValidator.cs (hunk 4 above)
Comment (bgavrilMS) [reply Extensions/AadIssuerValidator.cs]: I think our style is to prefer using 'var' when the type on right is apparent, which is the case here. The style of the samples does not have to match the style of the library, but it seems like a reasonable approach.

File: Extensions/AadIssuerValidator.cs [hunk 10]
```diff
@@ -23,29 +21,29 @@ public static class AadIssuerValidator
...
             if (jwtToken == null)
             {
                 throw new SecurityTokenInvalidIssuerException("Expecting a JWT Token from Azure Active Directory.");
             }
 
             // Extracting the tenant ID
-            string tenantId = jwtToken.Claims.FirstOrDefault(c => c.Type == "tid")?.Value;
+            var tenantId = jwtToken.Claims.FirstOrDefault(c => c.Type == "tid")?.Value;
```
Comment (bgavrilMS) [inline Extensions/AadIssuerValidator.cs]: In these cases I don't think the type is apparent and it would be a bit more readable to use string.

File: Extensions/AadIssuerValidator.cs (hunk 4 above)
Comment (bgavrilMS) [reply Extensions/AadIssuerValidator.cs]: Also, VS 2019 now proposes 3 rules to deal with var

1. Don't use var for standard types (e.g. int i = 3) - not sure if this extends to string, but it makes sense.
2. Use or Don't use var when type of the right is apparent (e.g. var x = new C() is ok, var y = obj.GetStuff() is not ok)
3. Use or Don't use var everywhere else.

File: Extensions/InMemoryTokenCacheHelper.cs [hunk 11]
```diff
@@ -58,20 +52,20 @@ public InMemoryTokenCacheProvider(IMemoryCache cache)
...
             return helper.GetMsalCacheInstance();
         }
     }
 
     public class InMemoryTokenCacheHelper
     {
-        string UserId = string.Empty;
-        string CacheId = string.Empty;
-        IMemoryCache memoryCache;
+        readonly string UserId = string.Empty;
```
Comment (bgavrilMS) [inline Extensions/InMemoryTokenCacheHelper.cs]: nit: add visibility (private)

File: Extensions/InMemoryTokenCacheHelper.cs [hunk 12]
```diff
@@ -103,7 +97,7 @@ public void Load()
         public void Persist()
         {
             // Reflect changes in the persistent store
-            byte[] blob = cache.Serialize();
+            var blob = cache.Serialize();
```
Comment (bgavrilMS) [inline Extensions/InMemoryTokenCacheHelper.cs]: not apparent, use byte[]

File: Startup.cs (hunk 9 above)
Comment (bojanv) [reply Startup.cs]: Let me refactor it and fix other suggestions / fixes as well. Will do it later this evening.

**Comment by bojanv** [comment]
Provided an update on reviewed code rules to unite them in the sample and removed some of the unnecessary boiler plate code to get instance of the types, which are already in place through built in dependency injection - please check and let me know //cc @snobu

**Comment by jmprieur** [comment]
FYI: working on merging this @bojanv

File: Infrastructure/AzureAdOptions.cs (hunk 2 above)
Comment (jmprieur) [reply Infrastructure/AzureAdOptions.cs]: @bojanv : I'm still not clear why you don't use AzureADOptions (provided by AspNet Core)

File: Infrastructure/Constants.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
+namespace WebApp_OpenIDConnect_DotNet.Infrastructure
+{
+    public static class Constants
+    {
+        public const string AdditionalClaims = "claims";
+        public const string OpenIdResponseType = "id_token code";
```
Comment (jmprieur) [inline Infrastructure/Constants.cs]: We might be able to use `OpenIdConnectResponseType.CodeIdToken`
(I need to check if the order matters)

File: Infrastructure/Constants.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+        public const string ScopeOfflineAccess = "offline_access";
+        public const string ScopeUserRead = "User.Read";
+        public const string ScopeProfile = "profile";
+        public const string ScopeOpenId = "openid";
+        public const string AuthorizationScheme = "Bearer";
```
Comment (jmprieur) [inline Infrastructure/Constants.cs]: BearerAuthorizationScheme?

File: Interfaces/IGraphApiOperations.cs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+using System.Threading.Tasks;
+
+namespace WebApp_OpenIDConnect_DotNet.Interfaces
+{
+    public interface IGraphApiOperations
+    {
+        Task<dynamic> CallOnBehalfOfUserAsync(string accessToken);
```
Comment (jmprieur) [inline Interfaces/IGraphApiOperations.cs]: CallMicrosoftGraphMeEndpoint

File: Startup.cs (hunk 9 above)
Comment (jmprieur) [reply Startup.cs]: @bojanv : did I miss the refactoring you were mentioning ^^

**Comment by jmprieur** [comment]
@bojanv closing this PR as we've integrated all your good ideas part of the [Tutorials](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/tutorials) branch.

Thanks for your contributions
